clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -850,47 +930,36 @@ synchronized boolean hasOngoingTransaction() {
-                // The UNKNOWN_PRODUCER_ID error code was added along with the new ProduceResponse which includes the
-                // logStartOffset. So the '-1' sentinel is not for backward compatibility. Instead, it is possible for
-                // a broker to not know the logStartOffset at when it is returning the response because the partition
-                // may have moved away from the broker from the time the error was initially raised to the time the
-                // response was being constructed. In these cases, we should just retry the request: we are guaranteed
-                // to eventually get a logStartOffset once things settle down.
-                return true;
-
-            if (batch.sequenceHasBeenReset()) {
-                // When the first inflight batch fails due to the truncation case, then the sequences of all the other
-                // in flight batches would have been restarted from the beginning. However, when those responses
-                // come back from the broker, they would also come with an UNKNOWN_PRODUCER_ID error. In this case, we should not
-                // reset the sequence numbers to the beginning.
+        // For an UNKNOWN_PRODUCER_ID, the idempotent producer will locally bump the epoch and reset the sequence
+        // numbers of in-flight batches from sequence 0, then retry the failed batch, which should now succeed.
+        // For the transactional producer, allow the batch to fail. When processing the failed batch, we will
+        // transition to an abortable error and set a flag indicating that we need to bump the epoch
+        if (error == Errors.UNKNOWN_PRODUCER_ID && !isTransactional()) {
+            bumpEpoch();
+            return true;
+        } else if (error == Errors.OUT_OF_ORDER_SEQUENCE_NUMBER) {
+            if (!hasUnresolvedSequence(batch.topicPartition) &&
+                    (batch.sequenceHasBeenReset() || !isNextSequence(batch.topicPartition, batch.baseSequence()))) {
+                // We should retry the OutOfOrderSequenceException if the batch is _not_ the next batch, ie. its base
+                // sequence isn't the lastAckedSequence + 1.
                 return true;
-            } else if (lastAckedOffset(batch.topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER) < response.logStartOffset) {
-                // The head of the log has been removed, probably due to the retention time elapsing. In this case,
-                // we expect to lose the producer state. Reset the sequences of all inflight batches to be from the beginning
-                // and retry them.
-                startSequencesAtBeginning(batch.topicPartition);
+            } else if (!isTransactional() && hasUnresolvedSequence(batch.topicPartition) &&
+                    isNextSequenceForUnresolvedPartition(batch.topicPartition, batch.baseSequence())) {
+                bumpEpoch();
+                this.partitionsWithUnresolvedSequences.clear();
```

Comment:
Here we are clearing the unresolved partitions, which may include partitions other than the one that is being retried. Should we do the same when the error is UNKNOWN_PRODUCER_ID?